### ISSUE TITLE
ReadNotify and EventAdd Responses do not contain PV information

### ISSUE THREAD

**Issue by vstadnytskyi** [caproto/caproto#545]
I have made one callback function that I call whenever a subscribed value changes its' value on a server. The idea is that I can just pass that information to a code that will use it and act on it if needed. In this way I can write one module that will act as input/output(IO) and simply pass the incoming information to the main system code(system. If something needs to be published the system sends information to IO and it takes care of it.  However, I discovered that ReadNotifyResponse a resulting object from read() and EventAddResponse a resulting object from subscribe() do not contain information about the PV. Hence, I cannot distinguish between callbacks generated by different PVs. 

Is there a reason why Response objects don't have PV information?

Examples:

If I subscribe to a PV and have a callback function that just prints the content of the response 
"self.sub.add_callback(self.monitor)"
```
def monitor(self,response):
        print(f'system monitor received {response}')

system monitor received EventAddResponse(data=array([2.37]), data_type=<ChannelType.DOUBLE: 6>, data_count=1, status=CAStatusCode(name='E
CA_NORMAL', code=0, code_with_severity=1, severity=<CASeverity.SUCCESS: 1>, success=1, defunct=False, description='Normal successful comp
letion'), subscriptionid=6, metadata=None)
```
and 
if I try to read the PV with sync client  "sync.client.read()"
```
pv.read()
Out[12]: ReadNotifyResponse(data=array([2.37]), data_type=<ChannelType.DOUBLE: 6>, data_count=1, status=CAStatusCode(name='ECA_NORMAL', c
ode=0, code_with_severity=1, severity=<CASeverity.SUCCESS: 1>, success=1, defunct=False, description='Normal successful completion'), ioi
d=7, metadata=None)

```

**Comment by vstadnytskyi** [comment]
Something else I found related to ReadNotifyResponse.
Somehow there are attributes with None values. Is it a bug or intentional and reserver for future expansion?
 
```
In [35]: resp = client.pvs['TOWER.P1.BUSY'].read()
In [36]: resp
Out[36]: ReadNotifyResponse(data=array([0], dtype=uint16), data_type=<ChannelType.ENUM: 3>, data_count=1, status=CAStatusCode(name='ECA
_NORMAL', code=0, code_with_severity=1, severity=<CASeverity.SUCCESS: 1>, success=1, defunct=False, description='Normal successful comp
letion'), ioid=11, metadata=None)

In [37]: type(resp.sender_address)
Out[37]: NoneType
```

**Comment by vstadnytskyi** [comment]
Can I look up the channel info(and its' PV) based on `subscriptionid`?

**Comment by danielballan** [comment]
The objects `ReadNotifyResponse`, `EventAddResponse`, and in general all the [commands](https://caproto.github.io/caproto/api.html#commands) are a literal representation of the Channel Access protocol. If you do `bytes(ReadNotifyResponse(...))` you'll get the bytes that were sent over the wire. Thus, these objects are designed to contain nothing more or less than what [the protocol specifies](https://epics.anl.gov/docs/CAproto.html).

In specific cases you mention, the protocol uses the `ioid` of the `ReadNotifyResponse` and the `subscriptionid` of the `EventAddResponse` to match a response back to the request that triggered it and, via that, the PV. You can see that happening here:

https://github.com/caproto/caproto/blob/334679b45fef52e998a4cbb398d95b5708b82c5b/caproto/threading/client.py#L1396-L1401

https://github.com/caproto/caproto/blob/334679b45fef52e998a4cbb398d95b5708b82c5b/caproto/threading/client.py#L1423-L1434

where `sub` is a `Subscription` object and has a `pv` attribute. That's how you could address your last question, looking up a PV by a `subscriptionid`.

Finally, this attribute

```
In [37]: type(resp.sender_address)
Out[37]: NoneType
```

is only used for certain commands---I believe the UDP ones.

**Comment by tacaswell** [comment]
I wonder if we should change the API on the non-pyepics compat subscription so that the signature of the callback also include the PV that it is acting on?

On one hand, I agree that we should not extend the Message objects beyond what is on the wire, on the other hand, doing the look up of subscriptionid -> PV on every callback is also annoying.

In the mean time you can do things like

```python
class XYZ:
    def monitor(self,response, PV):
        print(f'system monitor received {response} from {PV}')

    def some_function():
        self.sub.add_callback(functools.partialmethod(self.monitor, PV=self.sub))
```

to inject the PV into the callback at registration time.

**Comment by vstadnytskyi** [comment]
I do like the idea that "...The objects ReadNotifyResponse, EventAddResponse, and in general all the commands are a literal representation of the Channel Access protocol. ..." however, it would be convenient if one can extract the PV name from the subscriptionid.

Here how I am extracting PVname from subsciprption id
 
first, I save all subscriptions in self.sub dictionary.
```
self.pvs[pv], = ctx.get_pvs(pv)
self.sub[pv] = self.pvs[pv].subscribe()
```
where pv is the pv name I am subscribing to.

then I figured that from sub object I can extract both id and pvname and I am putting this pair in a separate lookup dictionary
```
id = self.sub[pv].subscriptionid
pv_name = self.sub[pv].pv.name
self.sub_dict[id] = pv_name
```

**Comment by danielballan** [comment]
Closed by #547 and #578
